File: LeetCode/Easy/0977-squares-of-a-sorted-array/0977-squares-of-a-sorted-array.java
class Solution {
    int i = 0;
    public int[] sortedSquares(int[] nums) {
        int[] result = new int[nums.length];
        int left = 0;
        int right = nums.length - 1;
        // 再定义一个指针，指向新数组的最后一个元素
        int finalIndex = nums.length - 1;
        while (left <= right) {
            int leftSquare = nums[left] * nums[left];
            int rightSquare = nums[right] * nums[right];
            if (leftSquare > rightSquare) {
                result[finalIndex] = leftSquare;
                left++;
            } else {
                result[finalIndex] = rightSquare;
                right--;
            }
            finalIndex--;

        }
        return result;
    }
}

File: LeetCode/Easy/0977-squares-of-a-sorted-array/0977-squares-of-a-sorted-array.py
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        n = len(nums)
        result = [0]*n
        left = 0
        right = n - 1
        p = n - 1
        while left <= right:
            if abs(nums[left]) > abs(nums[right]):
                result[p] = nums[left] * nums[left]
                left += 1
            else:
                result[p] = nums[right] * nums[right]
                right -= 1
            p -= 1
        return result
